File: core/tests/test_builtin_embedder.cpp
// Built-in local embedding inference: the zero-setup "just give it text" path.
//
// Tests five not-yet-implemented members of RetrievalEngine --
// load_embedding_model(), has_embedding_model(), embedding_dim(), embed(),
// add_text() and search_text() -- declared in retrieval_engine.hpp but
// with NO implementation anywhere yet. Building this test is expected to
// FAIL at link time with undefined-symbol errors: that is the correct,
// expected result for this pass (Phase 1, RED). Do NOT fix it by writing
// an implementation, a tokenizer, or a text-to-vector pipeline; that is
// Phase 2 (GREEN).
//
// Model file: rather than shipping a real ONNX/GGUF model into the test
// tree, these tests write a tiny *mock* model file (see WriteMockModel()
// and docs/DECISIONS.md for its documented contract). GREEN is expected to
// recognize that mock header and activate a deterministic, dependency-free
// test embedder (lexical-overlap cosine similarity) so the end-to-end text
// path can be exercised here without a multi-megabyte download. Loading a
// genuine ONNX/GGUF model is a separate, model-availability-gated path and
// is out of scope for this pass.
#include <gtest/gtest.h>

#include <cmath>
#include <cstddef>
#include <cstdio>
#include <filesystem>
#include <fstream>
#include <numeric>
#include <stdexcept>
#include <string>
#include <vector>

#include "retrieval_engine/retrieval_engine.hpp"

namespace {

// all-MiniLM-L6-v2's output width; also the dimension the mock model
// advertises. 768 (nomic-embed-text) would work equally well -- the tests
// only require embed()'s output to match this exactly.
constexpr std::size_t kEmbeddingDim = 384;

// Writes the mock embedding-model file the tests point load_embedding_model()
// at. Contract (also recorded in docs/DECISIONS.md): a file whose first
// line is exactly "RETRIEVAL_ENGINE_MOCK_EMBEDDING_MODEL v1" followed by a
// "dim=<n>" line selects a built-in deterministic test embedder instead of
// a real ONNX/GGUF runtime.
void WriteMockModel(const std::string& path, std::size_t dim) {
    std::ofstream out(path, std::ios::binary | std::ios::trunc);
    ASSERT_TRUE(out.is_open()) << "could not create mock model file at " << path;
    out << "RETRIEVAL_ENGINE_MOCK_EMBEDDING_MODEL v1\n";
    out << "dim=" << dim << "\n";
}

double CosineSimilarity(const std::vector<float>& a, const std::vector<float>& b) {
    EXPECT_EQ(a.size(), b.size());
    double dot = 0.0;
    double norm_a = 0.0;
    double norm_b = 0.0;
    for (std::size_t i = 0; i < a.size(); ++i) {
        dot += static_cast<double>(a[i]) * static_cast<double>(b[i]);
        norm_a += static_cast<double>(a[i]) * static_cast<double>(a[i]);
        norm_b += static_cast<double>(b[i]) * static_cast<double>(b[i]);
    }
    if (norm_a == 0.0 || norm_b == 0.0) {
        return 0.0;
    }
    return dot / (std::sqrt(norm_a) * std::sqrt(norm_b));
}

retrieval_engine::RetrievalEngine MakeEngineWithMockModel(const std::string& db_path,
                                                          const std::string& model_path) {
    std::remove(db_path.c_str());
    WriteMockModel(model_path, kEmbeddingDim);

    retrieval_engine::RetrievalEngine engine(db_path, kEmbeddingDim);
    engine.load_embedding_model(model_path);
    EXPECT_TRUE(engine.has_embedding_model());
    EXPECT_EQ(engine.embedding_dim(), kEmbeddingDim);
    return engine;
}

}  // namespace

// embed("Hello world") must return a non-empty vector of exactly the
// model's dimension.
TEST(BuiltinEmbedder, EmbedProducesNonEmptyVectorOfExactModelDimension) {
    retrieval_engine::RetrievalEngine engine =
        MakeEngineWithMockModel("builtin_embedder_embed.sqlite3", "builtin_embedder_mock_model.txt");

    const std::vector<float> embedding = engine.embed("Hello world");

    ASSERT_EQ(embedding.size(), kEmbeddingDim);
    const double magnitude = std::sqrt(std::inner_product(embedding.begin(), embedding.end(),
                                                          embedding.begin(), 0.0));
    EXPECT_GT(magnitude, 0.0) << "embedding must not be the all-zeros vector";
}

// The raw-text ingestion + query path: add_text() / search_text() with no
// caller-supplied float vectors anywhere. The document that shares
// vocabulary with the query must come back first.
TEST(BuiltinEmbedder, AddTextAndSearchTextRoundTripWithoutCallerSuppliedVectors) {
    retrieval_engine::RetrievalEngine engine =
        MakeEngineWithMockModel("builtin_embedder_search_text.sqlite3", "builtin_embedder_mock_model.txt");

    engine.add_text({
        retrieval_engine::TextDocumentInput{"install-guide",
                                            "python package installation guide for beginners", "", 0},
        retrieval_engine::TextDocumentInput{"weather-report",
                                            "mountain weather forecast cold and snowy today", "", 0},
        retrieval_engine::TextDocumentInput{"cooking-notes",
                                            "recipe roasted vegetables garlic thyme dinner", "", 0},
    });
    ASSERT_EQ(engine.chunk_count(), 3u);

    const std::vector<retrieval_engine::ChunkSearchResult> results =
        engine.search_text("the python package installation guide", /*k=*/2);

    ASSERT_FALSE(results.empty());
    EXPECT_EQ(results.front().document_id, "install-guide");
}

// Semantically related text (shared vocabulary) must have higher cosine
// similarity than unrelated text, end to end through embed().
TEST(BuiltinEmbedder, SemanticallyRelatedTextHasHigherCosineSimilarity) {
    retrieval_engine::RetrievalEngine engine =
        MakeEngineWithMockModel("builtin_embedder_cosine.sqlite3", "builtin_embedder_mock_model.txt");

    const std::vector<float> query = engine.embed("the python package installation guide");
    const std::vector<float> related = engine.embed("python package installation guide for beginners");
    const std::vector<float> unrelated = engine.embed("mountain weather forecast cold and snowy");

    const double related_similarity = CosineSimilarity(query, related);
    const double unrelated_similarity = CosineSimilarity(query, unrelated);

    EXPECT_GT(related_similarity, unrelated_similarity);
    EXPECT_GT(related_similarity, 0.5);
}

// embed() before any model is attached is a programming error, not a
// silent empty result.
TEST(BuiltinEmbedder, EmbedWithoutAModelThrows) {
    std::remove("builtin_embedder_no_model.sqlite3");
    retrieval_engine::RetrievalEngine engine("builtin_embedder_no_model.sqlite3", kEmbeddingDim);

    EXPECT_FALSE(engine.has_embedding_model());
    EXPECT_THROW(engine.embed("no model loaded"), std::logic_error);
}

// --- Model-path resolution / corrupted-file handling ---------------------

TEST(BuiltinEmbedder, LoadEmbeddingModelWithMissingFileThrowsRuntimeError) {
    std::remove("builtin_embedder_missing.sqlite3");
    retrieval_engine::RetrievalEngine engine("builtin_embedder_missing.sqlite3", kEmbeddingDim);

    EXPECT_THROW(engine.load_embedding_model("no_such_model_file_98765.bin"), std::runtime_error);
    EXPECT_FALSE(engine.has_embedding_model());
}

TEST(BuiltinEmbedder, LoadEmbeddingModelWithUnrecognizedFormatThrowsRuntimeError) {
    const std::string model_path = "builtin_embedder_bad_format.bin";
    {
        std::ofstream out(model_path, std::ios::binary | std::ios::trunc);
        out << "\x00\x01 this is not any embedding model format we know\n";
    }
    std::remove("builtin_embedder_badfmt.sqlite3");
    retrieval_engine::RetrievalEngine engine("builtin_embedder_badfmt.sqlite3", kEmbeddingDim);

    EXPECT_THROW(engine.load_embedding_model(model_path), std::runtime_error);
    EXPECT_FALSE(engine.has_embedding_model());
}

TEST(BuiltinEmbedder, LoadEmbeddingModelWithCorruptDimLineThrowsRuntimeError) {
    const std::string model_path = "builtin_embedder_corrupt_dim.txt";
    {
        std::ofstream out(model_path, std::ios::binary | std::ios::trunc);
        out << "RETRIEVAL_ENGINE_MOCK_EMBEDDING_MODEL v1\n";
        out << "dim=not-a-number\n";
    }
    std::remove("builtin_embedder_corruptdim.sqlite3");
    retrieval_engine::RetrievalEngine engine("builtin_embedder_corruptdim.sqlite3", kEmbeddingDim);

    EXPECT_THROW(engine.load_embedding_model(model_path), std::runtime_error);
    EXPECT_FALSE(engine.has_embedding_model());
}

TEST(BuiltinEmbedder, LoadEmbeddingModelWithDimensionMismatchThrowsInvalidArgument) {
    const std::string model_path = "builtin_embedder_dim_mismatch.txt";
    WriteMockModel(model_path, kEmbeddingDim + 1);  // model says 385, engine wants 384
    std::remove("builtin_embedder_dimmismatch.sqlite3");
    retrieval_engine::RetrievalEngine engine("builtin_embedder_dimmismatch.sqlite3", kEmbeddingDim);

    EXPECT_THROW(engine.load_embedding_model(model_path), std::invalid_argument);
    EXPECT_FALSE(engine.has_embedding_model());
}

// A ".onnx" model needs its vocab.txt beside it. Whether or not this build
// has the ONNX backend, pointing the loader at an .onnx file with no usable
// vocab is a std::runtime_error (never a crash, never silent).
TEST(BuiltinEmbedder, LoadOnnxModelWithoutAUsableSiblingVocabThrowsRuntimeError) {
    namespace fs = std::filesystem;
    const fs::path dir = "builtin_embedder_onnx_no_vocab_dir";
    fs::remove_all(dir);
    fs::create_directory(dir);
    {
        std::ofstream model(dir / "model.onnx", std::ios::binary | std::ios::trunc);
        model << "this is not a real ONNX protobuf";
    }

    std::remove("builtin_embedder_onnx_no_vocab.sqlite3");
    retrieval_engine::RetrievalEngine engine("builtin_embedder_onnx_no_vocab.sqlite3", kEmbeddingDim);

    EXPECT_THROW(engine.load_embedding_model((dir / "model.onnx").string()), std::runtime_error);
    EXPECT_FALSE(engine.has_embedding_model());

    fs::remove_all(dir);
}

// A corrupt .onnx file (vocab present) is likewise a clean std::runtime_error.
TEST(BuiltinEmbedder, LoadCorruptOnnxModelThrowsRuntimeError) {
    namespace fs = std::filesystem;
    const fs::path dir = "builtin_embedder_onnx_corrupt_dir";
    fs::remove_all(dir);
    fs::create_directory(dir);
    {
        std::ofstream model(dir / "model.onnx", std::ios::binary | std::ios::trunc);
        model << std::string("\x00\x01\x02 definitely not protobuf", 27);
    }
    {
        std::ofstream vocab(dir / "vocab.txt", std::ios::binary | std::ios::trunc);
        vocab << "[PAD]\n[UNK]\n[CLS]\n[SEP]\nhello\nworld\n";
    }

    std::remove("builtin_embedder_onnx_corrupt.sqlite3");
    retrieval_engine::RetrievalEngine engine("builtin_embedder_onnx_corrupt.sqlite3", kEmbeddingDim);

    EXPECT_THROW(engine.load_embedding_model((dir / "model.onnx").string()), std::runtime_error);
    EXPECT_FALSE(engine.has_embedding_model());

    fs::remove_all(dir);
}

// A failed re-load must not detach the model that was already working.
TEST(BuiltinEmbedder, LoadEmbeddingModelKeepsThePreviousModelWhenTheNewLoadFails) {
    retrieval_engine::RetrievalEngine engine =
        MakeEngineWithMockModel("builtin_embedder_rollback.sqlite3", "builtin_embedder_mock_model.txt");
    ASSERT_TRUE(engine.has_embedding_model());

    EXPECT_THROW(engine.load_embedding_model("no_such_model_file_98765.bin"), std::runtime_error);

    EXPECT_TRUE(engine.has_embedding_model());
    EXPECT_EQ(engine.embed("the model is still attached").size(), kEmbeddingDim);
}

File: tests/test_builtin_embedder.py
"""Built-in local embedding inference ("just give it text"), from Python.

Exercises the Python Engine wrapper and the CLI for:

    engine.load_embedding_model(model_path)
    engine.embed(text) -> list[float]
    engine.add_text(documents)          # documents: [{"id", "text", "metadata"}]
    engine.search_text(query_text, top_k)

The detailed embedding behavior (exact dimension, semantic-similarity
ordering, loader error handling) is covered more thoroughly in the C++
suites (core/tests/test_builtin_embedder.cpp, core/tests/test_onnx_embedder.cpp);
this file confirms the same capability is reachable from Python and the CLI.

Most tests use a tiny *mock* model file rather than a real ONNX/GGUF model
(see _write_mock_model() and docs/DECISIONS.md for the contract). The two
tests that need the real all-MiniLM-L6-v2 model skip when it is absent.
"""

import math
import subprocess
from pathlib import Path

import pytest

import retrieval_engine

_EMBEDDING_DIM = 384

_REPO_ROOT = Path(__file__).resolve().parent.parent
_ENGINE_CLI = _REPO_ROOT / ".venv" / "bin" / "engine"

_MINILM_DIM = 384


def _onnx_model_path() -> Path:
    """Where the real all-MiniLM-L6-v2 ONNX model is expected. The 90 MB
    file is not vendored, so the ONNX tests skip when it is absent.
    """
    import os

    from_env = os.environ.get("RETRIEVAL_ENGINE_TEST_ONNX_MODEL")
    if from_env:
        return Path(from_env)
    return Path.home() / ".cache" / "retrieval-engine" / "all-MiniLM-L6-v2" / "model.onnx"


def _write_mock_model(path) -> str:
    """Writes the mock embedding-model file (see module docstring / DECISIONS.md).

    First line selects a deterministic, dependency-free test embedder in
    place of a real ONNX/GGUF runtime.
    """
    path.write_text(
        f"RETRIEVAL_ENGINE_MOCK_EMBEDDING_MODEL v1\ndim={_EMBEDDING_DIM}\n",
        encoding="utf-8",
    )
    return str(path)


def _cosine(a, b) -> float:
    dot = sum(x * y for x, y in zip(a, b))
    norm_a = math.sqrt(sum(x * x for x in a))
    norm_b = math.sqrt(sum(y * y for y in b))
    if norm_a == 0.0 or norm_b == 0.0:
        return 0.0
    return dot / (norm_a * norm_b)


def _engine_with_mock_model(tmp_path, db_filename):
    model_path = _write_mock_model(tmp_path / "mock_model.txt")
    engine = retrieval_engine.Engine(str(tmp_path / db_filename), dim=_EMBEDDING_DIM)
    engine.load_embedding_model(model_path)
    return engine


def test_embed_returns_non_empty_vector_of_model_dimension(tmp_path):
    engine = _engine_with_mock_model(tmp_path, "embed.sqlite3")

    embedding = engine.embed("Hello world")

    assert len(embedding) == _EMBEDDING_DIM
    assert any(value != 0.0 for value in embedding)


def test_add_text_and_search_text_without_precomputed_vectors(tmp_path):
    engine = _engine_with_mock_model(tmp_path, "search_text.sqlite3")

    engine.add_text(
        [
            {"id": "install-guide", "text": "python package installation guide for beginners"},
            {"id": "weather-report", "text": "mountain weather forecast cold and snowy today"},
            {"id": "cooking-notes", "text": "recipe roasted vegetables garlic thyme dinner"},
        ]
    )
    assert engine.chunk_count() == 3

    results = engine.search_text("the python package installation guide", top_k=2)

    assert results
    assert results[0]["document_id"] == "install-guide"


def test_semantically_related_text_scores_higher(tmp_path):
    engine = _engine_with_mock_model(tmp_path, "cosine.sqlite3")

    query = engine.embed("the python package installation guide")
    related = engine.embed("python package installation guide for beginners")
    unrelated = engine.embed("mountain weather forecast cold and snowy")

    assert _cosine(query, related) > _cosine(query, unrelated)
    assert _cosine(query, related) > 0.5


def test_cli_routes_text_through_the_builtin_embedder(tmp_path):
    """`engine ingest/query --model <m> --model-dim <n>` generates vectors
    under the hood -- no caller-supplied floats at the CLI at all.
    """
    model_path = _write_mock_model(tmp_path / "mock_model.txt")

    docs_dir = tmp_path / "docs"
    docs_dir.mkdir()
    (docs_dir / "install.txt").write_text(
        "python package installation guide for beginners", encoding="utf-8"
    )
    (docs_dir / "weather.txt").write_text(
        "mountain weather forecast cold and snowy today", encoding="utf-8"
    )

    model_flags = ["--model", model_path, "--model-dim", str(_EMBEDDING_DIM)]

    ingest = subprocess.run(
        [str(_ENGINE_CLI), "ingest", str(docs_dir), *model_flags],
        capture_output=True,
        text=True,
        cwd=tmp_path,
    )
    assert ingest.returncode == 0, f"ingest failed: {ingest.stderr}"

    query = subprocess.run(
        [str(_ENGINE_CLI), "query", "the python package installation guide", *model_flags],
        capture_output=True,
        text=True,
        cwd=tmp_path,
    )
    assert query.returncode == 0, f"query failed: {query.stderr}"
    assert "install.txt" in query.stdout


def test_load_embedding_model_with_missing_file_raises(tmp_path):
    engine = retrieval_engine.Engine(str(tmp_path / "missing.sqlite3"), dim=_EMBEDDING_DIM)

    with pytest.raises(RuntimeError):
        engine.load_embedding_model(str(tmp_path / "does_not_exist.bin"))
    assert not engine.has_embedding_model()


def test_load_embedding_model_with_wrong_dimension_raises(tmp_path):
    # Mock model advertises a different dimension than the engine was built for.
    model_path = tmp_path / "wrong_dim_model.txt"
    model_path.write_text(
        f"RETRIEVAL_ENGINE_MOCK_EMBEDDING_MODEL v1\ndim={_EMBEDDING_DIM + 1}\n", encoding="utf-8"
    )
    engine = retrieval_engine.Engine(str(tmp_path / "wrong_dim.sqlite3"), dim=_EMBEDDING_DIM)

    with pytest.raises(ValueError):
        engine.load_embedding_model(str(model_path))
    assert not engine.has_embedding_model()


def _require_onnx_model():
    model = _onnx_model_path()
    if not model.exists():
        pytest.skip(f"all-MiniLM-L6-v2 ONNX model not found at {model}")
    return str(model)


def test_onnx_backend_embeds_and_retrieves_by_meaning(tmp_path):
    model_path = _require_onnx_model()
    engine = retrieval_engine.Engine(str(tmp_path / "onnx.sqlite3"), dim=_MINILM_DIM)
    engine.load_embedding_model(model_path)
    assert engine.has_embedding_model()
    assert engine.embedding_dim() == _MINILM_DIM

    vector = engine.embed("Where do I live?")
    assert len(vector) == _MINILM_DIM
    assert _cosine(vector, vector) == pytest.approx(1.0, abs=1e-5)  # unit length

    engine.add_text(
        [
            {"id": "home", "text": "I live in Berlin, the capital of Germany."},
            {"id": "weather", "text": "Heavy rain is expected across the region tomorrow."},
            {"id": "food", "text": "This pasta recipe calls for garlic and olive oil."},
        ]
    )
    results = engine.search_text("Which city is my house located in?", top_k=1)
    assert results
    assert results[0]["document_id"] == "home"


def test_onnx_backend_via_cli(tmp_path):
    model_path = _require_onnx_model()

    docs_dir = tmp_path / "docs"
    docs_dir.mkdir()
    (docs_dir / "home.txt").write_text(
        "I live in Berlin, the capital of Germany.", encoding="utf-8"
    )
    (docs_dir / "food.txt").write_text(
        "This pasta recipe calls for garlic and olive oil.", encoding="utf-8"
    )

    flags = ["--model", model_path, "--model-dim", str(_MINILM_DIM)]
    ingest = subprocess.run(
        [str(_ENGINE_CLI), "ingest", str(docs_dir), *flags],
        capture_output=True,
        text=True,
        cwd=tmp_path,
    )
    assert ingest.returncode == 0, f"ingest failed: {ingest.stderr}"

    query = subprocess.run(
        [str(_ENGINE_CLI), "query", "which city is my house in?", *flags],
        capture_output=True,
        text=True,
        cwd=tmp_path,
    )
    assert query.returncode == 0, f"query failed: {query.stderr}"
    assert "home.txt" in query.stdout


def test_cli_model_flag_requires_model_dim(tmp_path):
    docs_dir = tmp_path / "docs"
    docs_dir.mkdir()
    (docs_dir / "sample.txt").write_text("hello world", encoding="utf-8")

    result = subprocess.run(
        [str(_ENGINE_CLI), "ingest", str(docs_dir), "--model", "whatever.bin"],
        capture_output=True,
        text=True,
        cwd=tmp_path,
    )
    assert result.returncode == 1
    assert "--model-dim" in result.stderr
